moveos/moveos-wasm/src/wasm.rs

```diff
@@ -193,50 +236,62 @@ pub fn put_data_on_stack(instance: &mut WASMInstance, data: &[u8]) -> anyhow::Re
 }
 
 pub fn get_data_from_heap(
-    memory: Arc<Mutex<Memory>>,
+    memory: &mut Arc<Mutex<Memory>>,
     store: &Store,
     ptr_offset: i32,
-) -> anyhow::Result<Vec<u8>> {
-    let bindings = match memory.lock() {
-        Ok(v) => v,
-        Err(_) => return Err(anyhow::Error::msg("memory lock failed")),
-    };
+) -> Vec<u8> {
+    let bindings = memory.lock().expect("getting memory mutex failed");
     let memory_view = bindings.view(store);
     let mut length_bytes: [u8; 4] = [0; 4];
-    match memory_view.read(ptr_offset as u64, length_bytes.as_mut_slice()) {
-        Ok(_) => {}
-        Err(_) => return Err(anyhow::Error::msg("read memory failed")),
-    }
+    memory_view
+        .read(ptr_offset as u64, length_bytes.as_mut_slice())
+        .expect("read length_bytes failed");
     let length = u32::from_be_bytes(length_bytes);
     let mut data = vec![0; length as usize];
-    match memory_view.read((ptr_offset + 4) as u64, &mut data) {
-        Ok(_) => {}
-        Err(_) => return Err(anyhow::Error::msg("read memory failed")),
-    }
-    Ok(data)
-
-    // let ptr = memory_view.data_ptr().offset(ptr_offset as isize) as *mut c_char;
-    // let c_str = CStr::from_ptr(ptr);
-    // c_str.to_bytes().to_vec()
-    // let rust_str = c_str.to_str().expect("Bad encoding");
-    // let owned_str = rust_str.to_owned();
-    // owned_str
+    memory_view
+        .read((ptr_offset + 4) as u64, &mut data)
+        .expect("read uninit failed");
+    data
+}
+
+fn charge(env: FunctionEnvMut<Env>, amount: i64) -> Result<(), wasmer::RuntimeError> {
+    let mut gas_meter = env.data().gas_meter.lock().unwrap();
+    gas_meter.charge(amount as u64)
 }
 
-pub fn create_wasm_instance(bytecode: &Vec<u8>) -> anyhow::Result<WASMInstance> {
-    let mut store = Store::default();
-    let module = match Module::new(&store, bytecode) {
+pub fn create_wasm_instance(code: &[u8]) -> anyhow::Result<WASMInstance> {
+    // Create the GasMeter
+    let gas_meter = Arc::new(Mutex::new(GasMeter::new(GAS_LIMIT)));
+
+    // Create and configure the compiler
+    let mut compiler_config = wasmer::Cranelift::default();
```

Comment:
Using wasmer-compiler-singlepass instead of cranelift will improve compilation speed.

Reply:
OK, I'll test it.